src/Dompdf.php

```diff
@@ -502,37 +501,32 @@ public function loadHtml($str, $encoding = null)
             $quirksmode = false;
 
-            if ($this->options->isHtml5ParserEnabled() && class_exists(HTML5_Tokenizer::class)) {
-                $tokenizer = new HTML5_Tokenizer($str);
-                $tokenizer->parse();
-                $doc = $tokenizer->save();
-
-                $quirksmode = ($tokenizer->getTree()->getQuirksMode() > HTML5_TreeBuilder::NO_QUIRKS);
-            } else {
-                // loadHTML assumes ISO-8859-1 unless otherwise specified on the HTML document header.
-                // http://devzone.zend.com/1538/php-dom-xml-extension-encoding-processing/ (see #4)
-                // http://stackoverflow.com/a/11310258/264628
-                $doc = new DOMDocument("1.0", $encoding);
-                $doc->preserveWhiteSpace = true;
-                $doc->loadHTML($str);
-                $doc->encoding = $encoding;
-
-                // If some text is before the doctype, we are in quirksmode
-                if (preg_match("/^(.+)<!doctype/i", ltrim($str), $matches)) {
-                    $quirksmode = true;
-                } // If no doctype is provided, we are in quirksmode
-                elseif (!preg_match("/^<!doctype/i", ltrim($str), $matches)) {
-                    $quirksmode = true;
-                } else {
-                    // HTML5 <!DOCTYPE html>
-                    if (!$doc->doctype->publicId && !$doc->doctype->systemId) {
-                        $quirksmode = false;
-                    }
+            $dom = new DOMDocument("1.0", $encoding);
+            $dom->preserveWhiteSpace = true;
+            $html5 = new HTML5(['encoding' => $encoding, 'disable_html_ns' => true, 'target_document' => $dom]);
+            $html5->loadHTML($str);
+
+            // extra step to normalize the HTML document structure
+            // see Masterminds/html5-php#166
+            $doc = new DOMDocument("1.0", $encoding);
+            $doc->preserveWhiteSpace = true;
+            $doc->loadHtml($dom->saveHtml(), LIBXML_NOWARNING | LIBXML_NOERROR);
+
+            // If some text is before the doctype, we are in quirksmode
+            if (preg_match("/^(.+)<!doctype/i", ltrim($str), $matches)) {
+                $quirksmode = true;
+            } // If no doctype is provided, we are in quirksmode
+            elseif (!preg_match("/^<!doctype/i", ltrim($str), $matches)) {
+                $quirksmode = true;
+            } elseif ($doc->doctype !== null) {
+                // HTML5 <!DOCTYPE html>
+                if (!$doc->doctype->publicId && !$doc->doctype->systemId) {
+                    $quirksmode = false;
+                }
 
-                    // not XHTML
-                    if (!preg_match("/xhtml/i", $doc->doctype->publicId)) {
-                        $quirksmode = true;
-                    }
+                // not XHTML
+                elseif (!preg_match("/xhtml/i", $doc->doctype->publicId)) {
+                    $quirksmode = true;
```

Comment:
I did not catch that. Let me review that.

I was actually wondering about the value of quirksmode because right now the only place we use it, that I saw in a quick review, was in validating the content type of CSS when the value is false.

Reply:
Yes, that looks to be the only place. The original commit is https://github.com/dompdf/dompdf/commit/4d254ea6646995a28fc6ffa458366c65d1f1da00 for reference.

Is there a reason for using the `target_document` option instead of letting the parser create the `DOMDocument`?

Reply:
Honestly, I had a reason for doing so but now I don't recall what it was. I went through a few variations on the implementation and maybe that's left over from something I was trying out. Let me review and make sure I'm not forgetting a _good_ reason it's there.

Reply:
I may have done it that way to be able to set the preserveWhitespace property, as we did previously. However, the documentation for DOMDocument indicates that the [preserveWhitespace property](https://www.php.net/manual/en/class.domdocument.php#domdocument.props.preservewhitespace) should be true by default so this seems unnecessary. I'll continue to review and let you know if there's anything else.

Reply:
OK so I did some testing and I think supplying a DOMDocument instance is not necessary. I do think that was probably me accommodating the previous logic, even though it appears unnecessary.

Regarding doctype declarations and quicksmode...
html5-php uses the doctype during parsing, but the resulting output is always HTML5. Meaning that the HTML we're using will always be in standard mode. We don't do anything related to quirksmode during rendering anyway, and I don't see us doing any work in support of that (even though quirksmode [continues to live on](https://quirks.spec.whatwg.org/)). So perhaps we just remove the quirks detection logic entirely since with this change Dompdf always runs the document through the HTML5 parser?

Reply:
Yes, I think that makes sense.